libcontainer/container_linux.go

```diff
@@ -619,6 +619,33 @@ func (c *Container) newParentProcess(p *Process) (parentProcess, error) {
 	}
 
 	if safeExe != nil {
+		// Because we want to add safeExe to the set of ExtraFiles, if the fd of safeExe is
+		// too small, go stdlib will dup3 it to another fd, or dup3 a other fd to this fd,
+		// then it will cause the fd type cmd.Path refers to a random path. (#4294)
+
+		// +1 means safeExe.
+		minFd := stdioFdCount + len(cmd.ExtraFiles) + 1
+		if p.Init {
+			// This refers to fifo.
+			minFd++
+		}
+		if int(safeExe.Fd()) <= minFd {
+			maxFd, err := utils.GetMaxFds()
+			if err != nil {
+				return nil, fmt.Errorf("unable to get the max opened fd of current process: %w", err)
+			}
+			maxFd++
+			if err := unix.Dup3(int(safeExe.Fd()), maxFd, unix.O_CLOEXEC); err != nil {
+				return nil, fmt.Errorf("unable to dup3 the fd from %d to %d: %w", safeExe.Fd(), maxFd, err)
+			}
+			cmd.Path = "/proc/self/fd/" + strconv.Itoa(maxFd)
+			newSafeExe := os.NewFile(uintptr(maxFd), safeExe.Name())
+			if err := safeExe.Close(); err != nil {
+				return nil, fmt.Errorf("unable to close old safe exe(%d): %w", safeExe.Fd(), err)
+			}
+			safeExe = newSafeExe
```

Comment:
In fact, if we can ensure the safeExe's fd is bigger than the length of `cmd.ExtraFiles`, we could not have to add it it `cmd.ExtraFiles`.

Reply:
I think the child process doesn't need the fd refers to runc.

Reply:
From memory it needs to be at least 1 (or 2?) larger because Go will clobber files when setting up the files for execve. That is what https://go-review.googlesource.com/c/go/+/515799 was trying to fix.

The child process doesn't need it, but that was the only way to stop Go from clobbering it (though it turns out this doesn't work in this case).